k2/csrc/fsa_utils.cu

```diff
@@ -2149,4 +2150,453 @@ void FixNumStates(FsaVec *fsas) {
+       // we shouldn't have reached (has no arcs) or some other circumstance
+       // meant that no arc was chosen.
+       K2_DCHECK_LT(path_pos, num_batches);
+
+       g.sync();
+
+       FloatType p = shared_data->p;
+       int32_t begin_arc_idx01x = shared_data->begin_arc_idx01x,
+           num_arcs = shared_data->num_arcs;
+
+       if (num_arcs == 0) {
+         // If we reach a state with no arcs leaving it, it should be the final
+         // state; otherwise something went wrong.
+         K2_CHECK_EQ(fsas_row_splits2_data[final_state], begin_arc_idx01x);
+         if (thread_idx == 0) {
+           int32_t path_length = path_pos;
+           path_storage_start[0] = path_length;
+         }
+         return;
+       }
+
+       // Must sync again after read and before potential write..
+       g.sync();
+
+       int32_t arc_idx2 = thread_idx;
+       for (; arc_idx2 < num_arcs; arc_idx2 += g.size()) {
+         int32_t arc_idx012 = begin_arc_idx01x + arc_idx2;
+         FloatType interval_start = arc_cdf_data[arc_idx012],
+             interval_end = (arc_idx2 + 1 == num_arcs ? 1.0 :
+                             arc_cdf_data[arc_idx012 + 1]);
+         K2_DCHECK_GE(interval_end, interval_start);
+         K2_DCHECK_LE(interval_end, 1.0);
+         if (p >= interval_start && p <= interval_end &&
+             interval_end > interval_start &&
+             (p != interval_end || p == 1.0)) {
+           // The above if-statement ensures that any 0.0 <= p <= 1.0 will be
+           // inside exactly one interval.  The edges of the intervals form a
+           // non-decreasing sequence from 0.0 to 1.0.  So every such p is either
+           // completely inside an interval, or:
+           //    - Is 1.0 and is at the upper boundary of one nonempty interval
+           //    - Is 0.0 and is at the lower boundary of one nonempty interval
+           //    - Satisfies 0.0 < p < 1.0 and is at the boundary of two
+           //      nonempty intervals; we assign it to the higher of the two
+           //      intervals.
+           // The reason for the + 1 is that we'll put the path length in
+           // element 0.
+           path_storage_start[path_pos + 1] = arc_idx012;
+
+           // The following will ensure that 0.0 <= p <= 1.0: we know
+           // interval_end - interval_start > 0.0, p - interval_start >= 0.0,
+           // and (interval_end - interval_start) >= (p - interval_start).
+           p = (p - interval_start) / (interval_end - interval_start);
+
+           int32_t next_state_idx01 = arcs[arc_idx012].dest_state + state_idx0x,
+               next_arc_idx01x = fsas_row_splits2_data[next_state_idx01],
+               next_arc_idx01x_next = fsas_row_splits2_data[next_state_idx01 + 1];
+           shared_data->begin_arc_idx01x = next_arc_idx01x;
+           shared_data->num_arcs = next_arc_idx01x_next - next_arc_idx01x;
+           shared_data->p = p;
```

Comment:
Do you need a `break` here?

Reply:
Well spotted!  If I had checked my email more often I would have saved time debugging..